crates/containerd-shim-wasmtime/src/executor.rs

```diff
@@ -0,0 +1,133 @@
+use std::{fs::OpenOptions, os::fd::RawFd, path::PathBuf};
+
+use anyhow::{anyhow, Result};
+use containerd_shim_wasm::sandbox::oci;
+use libc::{dup, dup2, STDERR_FILENO, STDIN_FILENO, STDOUT_FILENO};
+use libcontainer::workload::{Executor, ExecutorError};
+use oci_spec::runtime::Spec;
+
+use wasmtime::{Engine, Linker, Module, Store};
+use wasmtime_wasi::WasiCtxBuilder;
+
+use crate::oci_wasmtime::{self, wasi_dir, wasi_file};
+
+const EXECUTOR_NAME: &str = "wasmtime";
+
+static mut STDIN_FD: Option<RawFd> = None;
+static mut STDOUT_FD: Option<RawFd> = None;
+static mut STDERR_FD: Option<RawFd> = None;
+
+pub struct WasmtimeExecutor {
+    pub stdin: Option<RawFd>,
+    pub stdout: Option<RawFd>,
+    pub stderr: Option<RawFd>,
+    pub engine: Engine,
+}
+
+impl Executor for WasmtimeExecutor {
+    fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
+        let args = oci::get_args(spec);
+        if args.is_empty() {
+            return Err(ExecutorError::InvalidArg);
+        }
+
+        let (mut store, f) = self
+            .prepare_function(spec, args)
+            .map_err(|err| ExecutorError::Other(format!("failed to prepare function: {}", err)))?;
+
+        log::info!("calling start function");
+        match f.call(&mut store, &[], &mut []) {
+            Ok(_) => std::process::exit(0),
+            Err(_) => std::process::exit(137),
+        };
+    }
+
+    fn can_handle(&self, _spec: &Spec) -> bool {
+        true
+    }
+
+    fn name(&self) -> &'static str {
+        EXECUTOR_NAME
+    }
+}
+
+impl WasmtimeExecutor {
+    fn prepare_function(
+        &self,
+        spec: &Spec,
+        args: &[String],
+    ) -> anyhow::Result<(Store<wasi_common::WasiCtx>, wasmtime::Func)> {
+        // already in the cgroup
+        let env = oci_wasmtime::env_to_wasi(spec);
+        log::info!("setting up wasi");
+
+        let path = wasi_dir(".", OpenOptions::new().read(true))?;
+        let mut wasi_builder = WasiCtxBuilder::new()
+            .args(args)?
+            .envs(env.as_slice())?
+            .preopened_dir(path, "/")?;
```

Comment:
This path contains a wasm binary. This might not be desired in the future. 

Major concenrs: james work, preload binary, and include static assets not included in the binary. 

rootfs
  - root
  - .wasm

Reply:
Note that here https://github.com/ipuustin/runwasi/commit/7e72c3ca10a0454d4e220aa35db26f710fb03a17#diff-36f92d3bdc22f6005e8d13ff459d9d135364b34748de944da89785ecaa8d9e0aR58 is how we could enable rootfs file access for WasmEdge shim too. We need to have a discussion what we want to do about it -- should we only expose /dev and /proc files to the runtime, or to the container too?

Reply:
I created an issue for this. Closing it now.